src/hubs/basehub.ts

```diff
@@ -0,0 +1,369 @@
+
+import { ColorDistanceSensor } from "../devices/colordistancesensor";
+import { CurrentSensor } from "../devices/currentsensor";
+import { Device } from "../devices/device";
+
+import { DuploTrainBaseColorSensor } from "../devices/duplotrainbasecolorsensor";
+import { DuploTrainBaseMotor } from "../devices/duplotrainbasemotor";
+import { DuploTrainBaseSpeaker } from "../devices/duplotrainbasespeaker";
+import { DuploTrainBaseSpeedometer } from "../devices/duplotrainbasespeedometer";
+
+import { HubLED } from "../devices/hubled";
+import { Light } from "../devices/light";
+import { MediumLinearMotor } from "../devices/mediumlinearmotor";
+import { MotionSensor } from "../devices/motionsensor";
+import { MoveHubMediumLinearMotor } from "../devices/movehubmediumlinearmotor";
+import { MoveHubTiltSensor } from "../devices/movehubtiltsensor";
+import { RemoteControlButton } from "../devices/remotecontrolbutton";
+import { SimpleMediumLinearMotor } from "../devices/simplemediumlinearmotor";
+import { PiezoBuzzer } from "../devices/piezobuzzer";
+import { TechnicLargeLinearMotor } from "../devices/techniclargelinearmotor";
+import { TechnicMediumHubAccelerometerSensor } from "../devices/technicmediumhubaccelerometersensor";
+import { TechnicMediumHubGyroSensor } from "../devices/technicmediumhubgyrosensor";
+import { TechnicMediumHubTiltSensor } from "../devices/technicmediumhubtiltsensor";
+import { TechnicXLargeLinearMotor } from "../devices/technicxlargelinearmotor";
+import { TiltSensor } from "../devices/tiltsensor";
+import { TrainMotor } from "../devices/trainmotor";
+import { VoltageSensor } from "../devices/voltagesensor";
+
+import * as Consts from "../consts";
+
+import Debug = require("debug");
+const debug = Debug("basehub");
+
+
+/**
+ * @class BaseHub
+ * @extends EventEmitter
+ */
+export class BaseHub extends EventEmitter {
+
+    protected _attachedDevices: {[portId: number]: Device} = {};
+    // protected _virtualPorts: {[portName: string]: Port} = {};
+
+    protected _name: string = "";
+    protected _firmwareVersion: string = "0.0.00.0000";
+    protected _hardwareVersion: string = "0.0.00.0000";
+    protected _primaryMACAddress: string = "00:00:00:00:00:00";
+    protected _batteryLevel: number = 100;
+    protected _rssi: number = -60;
+
+    protected _bleDevice: IBLEAbstraction;
+
+    private _type: Consts.HubType;
+    private _portMap: {[portName: string]: number} = {};
+    private _attachCallbacks: ((device: Device) => boolean)[] = [];
+
+    constructor (device: IBLEAbstraction, portMap: {[portName: string]: number} = {}, type: Consts.HubType = Consts.HubType.UNKNOWN) {
+        super();
+        this.setMaxListeners(20); // Technic Medium Hub has 9 built in devices + 4 external ports. Node.js throws a warning after 11 attached event listeners.
+        this._type = type;
+        this._bleDevice = device;
+        this._portMap = portMap;
+        device.on("disconnect", () => {
```

Comment:
I think it should be named `bleDevice` as for the attribute because `device.on("disconnect", () => {})` can have a very diferent meaning with the new paradigm